localstack/aws/protocol/parser.py

```diff
@@ -260,6 +260,59 @@ def _timestamp_unixtimestamp(timestamp_string: str) -> datetime.datetime:
     def _timestamp_rfc822(datetime_string: str) -> datetime.datetime:
         return parsedate_to_datetime(datetime_string)
 
+    @staticmethod
+    def _get_request_uri_regex(operation: OperationModel) -> Optional[Pattern[str]]:
+        """
+        Tries to extract the request URI from a given operation.
+        If a request URI can be found, it is transformed to a regular expression which can be used to try to match
+        incoming requests (i.e. their paths).
+
+        :param operation: to extract the requestUri and convert to a regex of
+        :return: extracted requestUri, converted to a regular expression
+        """
+        if operation is None:
+            return None
+        request_uri_regex = None
+        http = operation.http
+        if len(http) > 0:
+            request_uri = http.get("requestUri")
+            request_uri_regex = RequestParser._convert_request_uri_to_regex(request_uri)
+        return request_uri_regex
+
+    @staticmethod
+    def _convert_request_uri_to_regex(request_uri: str) -> Optional[Pattern[str]]:
+        """
+        Converts the given request_uri to a regular expression.
+        It adds fences (line start and line end operators) and replaces the path param placeholders (f.e. "{VarName}")
+        with named capture groups (where the capture group name is the variable placeholder name).
+
+        :param request_uri: to convert to a regex
+        :return: regex
+        """
+        if request_uri is None:
+            return None
+        # Replace the variable place holder (f.e. {ConnectionId} in "/fuu/{ConnectionId}/bar/")
+        # with a named regex group.
+        request_uri_regex = re.sub(
+            "{(?P<VariableName>[^/]*)}", "(?P<\\g<VariableName>>[^/]*)", request_uri
+        )
+        # Put regex in fences to make sure that we do not match any substrings
+        request_uri_regex = f"^{request_uri_regex}$"
+        # The result is a regex itself.
+        return re.compile(request_uri_regex)
```

Comment:
No action needed for this PR, but just to get your thoughts:
Would it be possible to implement this in a simpler way? simply breaking up each path component by `/` and then checking each component for `{`, then parsing that?

I understand that the regex approach is shorter in terms of code, but I have some concerns about performance, since we're running two `re` calls for every request, which are probably avoidable?

Reply:
We could give it a try, but this might be quite complex. This function does not perform any checks, it just prepares the request URI in the specification (f.e. `/2021-01-01/opensearch/domain/{DomainName}` for `delete_domain`) to be later matched against the incoming request.
This is necessary in order to determine which operation the request belongs to:
https://github.com/localstack/localstack/blob/86d48493931b810b8408dd8074b34a9d51d1f848/localstack/aws/protocol/parser.py#L518-L530
Afterwards we need the named capture groups in the regex to extract the path parameters from the path of the incoming request.

So, I definitely see the potential for the performance implication (especially about all the matches with the lookup using `next(filter(...))` in the snippet above), but the regular expressions are quite heavily used here. I am quite confident that we could increase the performance here by optimizing the path matching algorithm, but it seemed like an overkill for the first iteration.